pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1697,21 +1717,14 @@ private void sendSegmentRefreshMessage(OfflineSegmentZKMetadata segmentZKMetadat
    * @param segmentMetadata Meta-data for the segment, used to access segmentName and tableName.
    */
   // NOTE: method should be thread-safe
-  private void addNewOfflineSegment(SegmentMetadata segmentMetadata) {
+  private void addNewOfflineSegment(SegmentMetadata segmentMetadata, List<String> assignedInstances) {
```

Comment:
Should we allow null assignedInstances here? That might cause the problem we are trying to solve now

Reply:
Yes, the variable `assignedInstances` can be null, but it can't be **empty**. 
If it's empty, it's guaranteed that the method `getAssignedInstancesForSegment` gets called but there's no assigned instance. In this case it has been handled in `SegmentValidator` before hitting this part, i.e. throwing an exception in validator.
If it's null, the caller might be from `PerfBenchmarkDriver`. And it can also be that this is an existing segment, which shouldn't be in Method `addNewOfflineSegment`. So we're good here.

Reply:
This will only be called when it's a new segment. You can call getAssignedInstancesForSegment() when assignedInstances are not provided in addNewSegment(), but it should not be null here. I would recommend add a Preconditions.checkNotNull() because when we call this, the segment ZK metadata has been updated.

Reply:
`@Nonnull` has been added in front of the param.
And `getAssignedInstancesForSegment()` is called when `assignedInstances` not provided in Method `addNewSegment(@Nonnull SegmentMetadata segmentMetadata, @Nonnull String downloadUrl)`

Reply:
@Jackie-Jiang A Precondition check on `assignedInstances` is added before adding metadata to zk.